### DISCUSSION TITLE
Current approach to work with trailing slashes

### DISCUSSION THREAD

**Discussion by hoefling** [aio-libs/yarl#1156] [Q&A] [answered]
Hi, what is the best (i.e. intended) approach to add and drop trailing slashes? I am currently using
```python
>>> u = yarl.URL('https://www.example.com/spam/eggs')
>>> u
URL('https://www.example.com/spam/eggs')
>>> u / ''
URL('https://www.example.com/spam/eggs/')
```
for adding and
```python
>>> u = yarl.URL('https://www.example.com/spam/eggs/')
>>> u
URL('https://www.example.com/spam/eggs/')
>>> u.with_path(u.path.removesuffix('/'))
URL('https://www.example.com/spam/eggs')
```
for dropping the slashes, but maybe somebody has a more elegant solution, esp. for the dropping part.

**Comment by webknjaz** [comment]
@commonism might have opinions on this ^

**Reply by commonism** [reply, marked as answer]
for adding I don't think it'll get any better than 
```python
u / ''
```
which is even safe, as it does not add a / if the path ends with / already 
but could be improved to avoid unnecessary construction
```python
u / '' if u.parts[-1] != "" else u
```

removing / - possible via
```python
(u / "").parent
```
or - avoiding the cost of construction if not required: 
```python
u.parent if u.parts[-1] == "" else u
```

but it takes a certain angle to recognize elegance in this I guess

**Reply by hoefling** [reply]
@commonism uh I like the `(u / "").parent` one :+1:

**Reply by hoefling** [reply]
@commonism disregard that, `(u / "").parent` does not remove trailing slash from host.
```py
>>> u = yarl.URL('https://www.example.com/')
>>> u.parent
URL('https://www.example.com/')
```

**Reply by commonism** [reply]
This is partial due to the requirements of [RFC3986 - 6.2.3 - Scheme-Based Normalization](https://datatracker.ietf.org/doc/html/rfc3986.html#section-6.2.3)

```
   In general, a URI that uses the generic syntax for authority with an
   empty path should be normalized to a path of "/".
```

[Currently](https://github.com/aio-libs/yarl/pull/1023#discussion_r1636861213) this is incomplete in yarl as
```python
>>> yarl.URL("http://www.example.org")
URL('http://www.example.org')
```
does not add it.

Therefore currently 
```python
(u / '').parent if u.path != "/" else u.with_path("")
```
will do, but it will break once scheme based normalization is complete in yarl.

On the other hand - there is no URL without trailing / in http, maybe completing scheme-based normalization will do as well?
